File: node/parse.js
const readStdin = require('./stdin');

const DAY_PATTERN = '\\d+ день';
const RATIO_EXERCISE_PATTERN = '(\\d+)% (\\d+)х(\\d+)';
const SIMPLE_EXERCISE_PATTERN = '(\\d+)х(\\d+)';

function extractExercise(line, patterns) {
    return extractExerciseByRatioPattern(line, patterns) || extractExerciseBySimplePattern(line);
}

function checkPattern(line, pattern) {
    const check = new RegExp(pattern).exec(line);
    return check ? line.substr(0, check.index).trim() : null;
}

function selectWeightByName(name, patterns) {
    const obj = patterns.find(item => item.pattern.test(name));
    return obj ? obj.weight : null;
}

function calculateWeight(ratio, weight) {
    const val = ratio * weight;
    return Math.round(val / 5) * 5;
}

function extractExerciseByRatioPattern(line, patterns) {
    const name = checkPattern(line, RATIO_EXERCISE_PATTERN);
    if (!name) {
        return null;
    }
    const regexp = new RegExp(RATIO_EXERCISE_PATTERN, 'g');
    const weight = selectWeightByName(name, patterns);
    let item;
    const sets = [];
    while (item = regexp.exec(line)) {
        const ratio = Number(item[1]) / 100;
        const obj = {
            ratio,
            weight: calculateWeight(ratio, weight),
            count: Number(item[2])
        };
        const ii = Number(item[3]);
        for (let i = 0; i < ii; ++i) {
            sets.push(obj);
        }
    }
    return {
        name,
        sets
    };
}

function extractExerciseBySimplePattern(line) {
    const name = checkPattern(line, SIMPLE_EXERCISE_PATTERN);
    if (!name) {
        return null;
    }
    const item = new RegExp(SIMPLE_EXERCISE_PATTERN).exec(line);
    const obj = {
        count: Number(item[1])
    };
    const ii = Number(item[2]);
    const sets = [];
    for (let i = 0; i < ii; ++i) {
        sets.push(obj);
    }
    return {
        name,
        sets
    };
}

function parse(data) {
    const lines = data.split('\n').map(x => x.trim()).filter(x => x);
    const week = lines[0];
    const list = [];
    const regexp = new RegExp(DAY_PATTERN);
    lines.forEach((line, i) => {
        if (regexp.test(line)) {
            list.push(i);
        }
    });
    const patterns = [];
    const weights = [];
    lines.slice(1, list[0]).forEach((value) => {
        const parts = value.trim().split(' ');
        const name = parts[0].trim();
        const weight = Number(parts[1].trim());
        weights.push({ name, weight });
        patterns.push({
            pattern: new RegExp(name.toLowerCase(), 'i'),
            weight
        });
    });
    const days = list.map((pos, i) => {
        const items = lines.slice(pos + 1, list[i + 1]);
        return {
            name: lines[pos],
            exercises: items.map(line => extractExercise(line.trim(), patterns))
        };
    });
    return {
        week,
        weights,
        days
    };
}

readStdin().then((data) => {
    process.stdout.write(JSON.stringify(parse(data), null, 2) + '\n');
});

File: parse.py
#!/usr/bin/python3

import re
import json
from process import process

DAY_PATTERN = '\\d+ день'
RATIO_EXERCISE_PATTERN = re.compile('(\\d+)% (\\d+)х(\\d+)')
SIMPLE_EXERCISE_PATTERN = re.compile('(\\d+)х(\\d+)')

def parse(data):
    lines = re.split('\n+', data)

    lines = list(filter(None, map(str.strip, lines)))
    week = lines[0]

    positions = []
    for i, line in enumerate(lines):
        if re.search(DAY_PATTERN, line):
            positions.append(i)
    positions.append(len(lines))

    patterns = []
    weights = []
    for line in lines[1:positions[0]]:
        parts = line.split(' ')
        name = parts[0].strip()
        weight = int(parts[1].strip())
        pattern = re.compile(name.lower(), re.IGNORECASE)
        weights.append({ 'name': name, 'weight': weight })
        patterns.append({ 'pattern': pattern, 'weight': weight })

    days = []
    process_line = lambda line: extract_exercise(line, patterns)
    for i, pos in enumerate(positions[:-1]):
        items = lines[pos + 1:positions[i + 1]]
        days.append({
            'name': lines[pos],
            'exercises': list(map(process_line, items))
        })

    return json.dumps({
        'week': week,
        'weights': weights,
        'days': days
    }, ensure_ascii=False, indent=2) + '\n'

def extract_exercise(line, patterns):
    return extract_exercise_by_ratio_pattern(line, patterns) \
        or extract_exercise_by_simple_pattern(line)

def extract_exercise_by_ratio_pattern(line, patterns):
    name = check_pattern(line, RATIO_EXERCISE_PATTERN)
    if not name:
        return None
    weight = select_weight_by_name(name, patterns)
    sets = []
    for match in re.finditer(RATIO_EXERCISE_PATTERN, line):
        ratio, reps, count = match.groups()
        k = int(ratio) / 100
        obj = {
            'ratio': k,
            'weight': calculate_weight(k, weight),
            'count': int(reps)
        }
        for _ in range(int(count)):
            sets.append(obj)
    return {
        'name': name,
        'sets': sets
    }

def check_pattern(line, pattern):
    match = re.search(pattern, line)
    if match:
        return line[:match.start()].strip()
    return None

def select_weight_by_name(name, patterns):
    for obj in patterns:
        if obj['pattern'].search(name):
            return obj['weight']
    return None

def calculate_weight(ratio, weight):
    val = ratio * weight
    return round(val / 5) * 5

def extract_exercise_by_simple_pattern(line):
    name = check_pattern(line, SIMPLE_EXERCISE_PATTERN)
    if not name:
        return None
    count, reps = re.search(SIMPLE_EXERCISE_PATTERN, line).groups()
    sets = []
    obj = {
        'count': int(reps)
    }
    for _ in range(int(count)):
        sets.append(obj)
    return {
        'name': name,
        'sets': sets
    }

process(parse)
